sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,354 @@
+    )
+    setattr(app.builder.env, "sphinx_immaterial_custom_icons", {})
+    user_admonitions = []
+    for config_val in custom_admonitions:
+
+        admonition = CustomAdmonitionConfig(**dict(config_val))
+        app.add_directive(
+            name=admonition.name,
+            cls=get_directive_class(admonition.name, admonition.title),
+            override=admonition.override,
+        )
+
+        # set variables for CSS template to match HTML output from generated directives
+        admonition.name = nodes.make_id(admonition.name)
+        admonition.icon = load_svg_into_builder_env(app.builder, admonition.icon)
+        user_admonitions.append(admonition)
+    setattr(app.builder.env, "sphinx_immaterial_custom_admonitions", user_admonitions)
+
+
+def on_config_inited(app: Sphinx, config: Config):
+    """Add admonitions based on CSS classes inherited from mkdocs-material theme."""
+
+    # override the generic admonition directive
+    app.add_directive("admonition", get_directive_class("admonition", ""), True)
+
+    # generate directives for inherited admonitions from upstream CSS
+    if getattr(config, "sphinx_immaterial_generate_inherited_admonitions"):
+        for admonition in (
+            "abstract",
+            "info",
+            "success",
+            "question",
+            "failure",
+            "bug",
+            "example",
+            "quote",
+        ):
+            app.add_directive(
+                f"si-{admonition}",
+                get_directive_class(admonition, _(admonition.title())),
+            )
+
+    user_defined_admonitions = getattr(config, "sphinx_immaterial_custom_admonitions")
+    user_defined_dir_names = [
+        directive["name"] for directive in user_defined_admonitions
+    ]
+
+    # override the specific admonitions defined in sphinx and docutils
+    # these are the admonitions that have translated titles in sphinx.locale
+    for admonition, title in admonitionlabels.items():
+
+        if admonition in user_defined_dir_names:
+            continue
+        app.add_directive(admonition, get_directive_class(admonition, title), True)
+
+
+def consolidate_css(app: Sphinx, env: BuildEnvironment):
+    """Generates the CSS for icons and admonitions, then appends that to the
+    theme's bundled CSS."""
+
+    theme_options = app.config["html_theme_options"]
+    is_palette_defined = False
+    if theme_options:
+        is_palette_defined = "palette" in theme_options
+
+    custom_admonitions = getattr(env, "sphinx_immaterial_custom_admonitions")
+    custom_icons = getattr(env, "sphinx_immaterial_custom_icons")
+    jinja_env = jinja2.Environment(
+        loader=jinja2.FileSystemLoader(str(PurePath(__file__).parent))
+    )
+    template = jinja_env.get_template("custom_admonitions.css")
+    generated = template.render(
+        icons=custom_icons,
+        admonitions=custom_admonitions,
+    )
+
+    css_name = "stylesheets/sphinx_immaterial_theme.min.css"
```

Comment:
We could include a hash of the contents in the filename as in mkdocs-material --- the purpose of this is to allow users to mark these bundle files as cachable with a long TTL.  I'm not sure how useful that is in practice, but it isn't hard for us to.

To do that, instead of writing directly to the output file, you can first write to an `io.BytesIO()` object.  Then you can use `hashlib.sha256(out.getvalue()).hexdigest()[:16]` to compute a hash and use the first 16 hex digits (8 bytes) of it.

Reply:
I noticed that there will be multiple CSS files written to the output if re-building the docs with a change to the custom admonitions config. I think the expected behavior is to remove the outdated generated CSS files in this case.

Reply:
This is a general problem with Sphinx's build mechanism when not using a fresh output directory for each build invocation --- stale files from a prior build may remain.  For example, if you delete `xxx.rst`, `xxx.html` may remain in the output directory from a prior build.

We could add special logic to delete the stale CSS files, but I'm not sure it is particularly useful to do so, given that stale files can still remain for other reasons.

For the apigen modules, this problem is more severe because due to sphinx limitations we have to write files to the source directory, not just the output directory.  In that case we do have special logic to delete the stale files, but it is also a bit more trouble to get a fresh source directory for every build, than to get a fresh output directory for every build.  Additionally, the generated files in the source directory get processed by sphinx and therefore impact the result, while extra files in the output directory just get ignored.

Reply:
Yeah, it's not a big concern since the problem would only occur when connective builds have a change in the amount of `si-icon`s or a change in custom admonitions. In practice I think most people would be hosting docs that were made from a fresh build.